capnp/lib/capnp.pyx

```diff
@@ -3739,6 +3760,36 @@ cdef class _FlatArrayMessageReader(_MessageReader):
 
 
 @cython.internal
+cdef class _SegmentArrayMessageReader(_MessageReader):
+
+    cdef object _objects_to_pin
+
+    def __init__(self, segments):
+        # takes a Python array of bytes and constructs a ConstWordArrayArrayPtr
+        num_segments = len(segments)
+        cdef char* ptr
+        cdef schema_cpp.ConstWordArrayPtr seg_ptr
+        cdef schema_cpp.ConstWordArrayPtr* seg_ptrs = <schema_cpp.ConstWordArrayPtr*>malloc(num_segments * sizeof(schema_cpp.ConstWordArrayPtr))
+        self._objects_to_pin = []
+        for i in range(0, num_segments):
+            segment = bytes(segments[i])
```

Comment:
You should probably use `PyObject_AsReadBuffer` here (see https://github.com/jparyani/pycapnp/blob/032713f8b9120afc31642469560f38a49e539bf2/capnp/lib/capnp.pyx#L3450 for an example).